components/micro-gateway-core/src/main/ballerina/src/gateway/handler_providers/jwt_auth_provider.bal

```diff
@@ -97,11 +106,37 @@ public type JwtAuthProvider object {
                         } else {
                             printDebug(KEY_JWT_AUTH_PROVIDER, "jti claim not found in the jwt");
                         }
+                        var jwtTokenClaimCached = self.gatewayCache.retrieveClaimMappingCache(jwtToken);
+                        if (jwtTokenClaimCached is runtime:Principal) {
+                            invocationContext.principal =  jwtTokenClaimCached;
+                            printDebug(KEY_JWT_AUTH_PROVIDER, "Moddified claims in the cache");
+                        } else {
+                            printDebug(KEY_JWT_AUTH_PROVIDER, "Moddified claims is not in the cache");
+                            var result = doMappingContext(invocationContext, self.className, self.claims,
+                                self.classLoaded, jwtPayloadFromCache, self.jwtValidatorConfig, self.gatewayCache, authContext);
+                            if (result is auth:Error){
+                                return result;
+                            }
+                            jwtToken = authContext?.authToken.toString();
+                        }
                         return validateSubscriptions(jwtToken, cachedJwt.jwtPayload, self.subscriptionValEnabled, isGRPC);
-                    } 
+                    }
                     printDebug(KEY_JWT_AUTH_PROVIDER, "jwt not found in the jwt cache");
                     (jwt:JwtPayload | error) payload = getDecodedJWTPayload(jwtToken);
                     if (payload is jwt:JwtPayload) {
+                        var jwtTokenClaimCached = self.gatewayCache.retrieveClaimMappingCache(jwtToken);
```

Comment:
If there are no claims and additional java class specified then we don not always check the cache and invoke doMappingContext method